src/js/carousel.js

```diff
@@ -0,0 +1,114 @@
+(function (window) {
+  function createDots() {
+    var existingDots = this.$carousel.getElementsByClassName('dots');
+    if (existingDots.length) {
+      this.$carousel.removeChild(existingDots[0]);
+    }
+    var dots = document.createElement('ul');
+    dots.className = 'dots';
+    var items = this.$carousel.getElementsByTagName('ul')[0];
+    this.$count = items.childElementCount;
+    for (var i = 0; i < items.childElementCount; i += 1) {
+      var dot = document.createElement('li');
+      dot.dataset.index = i;
+      dots.appendChild(dot);
+    }
+    this.$carousel.appendChild(dots);
+  }
+
+  function selectView(index) {
+    var scrollPosition = index * this.$width;
+    var items = this.$carousel.firstElementChild;
+    items.scrollTo(scrollPosition, 0);
+  }
+
+  function unselectAllDots() {
+    var dots = this.$carousel.lastElementChild;
+    for (var i = 0; i < dots.childElementCount; i += 1) {
+      dots.children[i].className = '';
+    }
+  }
+
+  function highlightDot(index) {
+    this.unselectAllDots();
+    var dots = this.$carousel.lastElementChild;
+    for (var i = 0; i < dots.childElementCount; i += 1) {
+      if (i === index) {
+        dot = dots.children[i].className = 'active';
+      }
+    }
+  }
+
+  function update() {
+    this.selectView(this.$selectedView);
+    this.highlightDot(this.$selectedView);
+  }
+
+  function _onDotClick(e) {
+    if (e.target.tagName !== 'LI') return;
+    clearInterval(this.$interval);
+    this.$selectedView = parseInt(e.target.dataset.index, 10);
+    this.update();
+  }
+
+  function registerListeners() {
+    var dots = this.$carousel.lastElementChild;
+    dots.addEventListener('click', _onDotClick.bind(this));
+  }
+
+  function updateWidth() {
+    var width = this.$width.toString() + 'px';
+    this.$carousel.style.width = width;
+    var list = this.$carousel.querySelector('.items ul');
```

Comment:
Now I finally see a `.querySelector` here. Why you're using it here and have used `.getElementsByXXX` in other places?

Reply:
I prefer to use `getElementByXXX` functions because they are [better in performance](https://jsperf.com/getelementbyid-vs-queryselector). Also they are better supported by old browsers. I think I used `querySelector` here for convenience. No special reason, though.